File: nc-150/1D-DP/min_cost_climbing_stairs.cpp
#include <vector>

using namespace std;

/*
 * Author: Valerie Doan
 * Date: October 11, 2024
 * Problem: Min Cost Climbing Stairs
 * Level: Easy
 * Neetcode Course: Neetcode 150
 * Neetcode Topic: 1-D DP
 */

int minCostClimbingStairs(vector<int> &cost)
{
    vector<int> mem(cost.size(), 0);

    mem[0] = cost[0];
    mem[1] = cost[1];

    for (int i = 2; i < cost.size(); i++)
    {
        mem[i] = cost[i] + min(mem[i - 1], mem[i - 2]);
    }

    return min(mem[cost.size() - 1], mem[cost.size() - 2]);
}

File: python/nc-150/1D-DP/min_cost_climbing_stairs.py
"""
Author: Valerie Doan
Date: November 11, 2024
Problem: Min Cost Climbing Stairs
Level: Easy
Neetcode Course: Neetcode 150
Neetcode Topic: 1-D DP
"""


class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp = []

        if len(cost) == 1:
            return cost[0]

        dp.extend([cost[0], cost[1]])

        for i in range(2, len(cost)):
            dp.append(min(cost[i] + dp[i - 1], cost[i] + dp[i - 2]))

        return min(dp[len(dp) - 1], dp[len(dp) - 2])
